# Top 5 Queries

# Top 1 - Chapter 7 Exercise 1

- Write a query against the dbo.Orders table that computes for each customer order, both a rank and a dense rank, partitioned by custid, ordered by qty
- Tables Involved: dbo.Orders(Substituted with the new table created in the chapter)

Desired output:

| custid | orderid | qty | rnk | drnk |
| --- | --- | --- | --- | --- |
| A | 30001 | 10 | 1 | 1 |
| A | 40005 | 10 | 1 | 1 |
| A | 10001 | 12 | 3 | 2 |
| A | 40001 | 40 | 4 | 3 |
| B | 20001 | 12 | 1 | 1 |
| B | 30003 | 15 | 2 | 2 |
| B | 10005 | 20 | 3 | 3 |
| C | 10006 | 14 | 1 | 1 |
| C | 20002 | 20 | 2 | 2 |
| C | 30004 | 22 | 3 | 3 |
| D | 30007 | 30 | 1 | 1 |

In [8]:
use Northwinds2022TSQLV7
select custid, orderid, qty, 
rank() over (partition by custid order by qty) as rnk, 
DENSE_RANK() over (partition by custid order by qty) as drnk 
from dbo.Orders

(11 rows affected)

Total execution time: 00:00:00.042

custid orderid qty rnk drnk A 30001 10 1 1 A 40005 10 1 1 A 10001 12 3 2 A 40001 40 4 3 B 20001 12 1 1 B 30003 15 2 2 B 10005 20 3 3 C 10006 14 1 1 C 20002 20 2 2 C 30004 22 3 3 D 30007 30 1 1

### Explanation:

The proposition is asking us to use the dbo.orders table that we created earlier in the chapter. I placed the new table inside the northwinds database to have consistency. I select the custid, orderid, and qty from that table. The proposition outlines the necesities for rank and denserank using the partition by clause on the custid and ordered by the qty for both columns. The rank represents how many different orders have a larger quantity while the dense rank column represents the number of orders with a distinct value larger than current row's qty value. For both columns they add one becuase it shows the rank of each row starting from 1.

# Top 2 - Chapter 7 Exercise 2

- The following query against the Sales.OrderValues view returns distinct values and their associated row numbers

```
SELECT val, ROW_NUMBER() OVER(ORDER BY val) AS rownum
FROM Sales.OrderValues
GROUP BY val;
```

- Can you think of an alternative way to achieve the same task?

Tables involved: Sales.OrderValues view

\-- Desired output:

| val | rownum |
| --- | --- |
| 12.50 | 1 |
| 18.40 | 2 |
| 23.80 | 3 |
| 28.00 | 4 |
| 30.00 | 5 |
| 33.75 | 6 |
| 36.00 | 7 |
| 40.00 | 8 |
| 45.00 | 9 |
| 48.00 | 10 |
| ... |  |
| 12615.05 | 793 |
| 15810.00 | 794 |
| 16387.50 | 795 |

(795 row(s) affected)

In [9]:
use TSQLV6
;WITH temp AS
(SELECT distinct val FROM Sales.OrderValues)
SELECT val, ROW_NUMBER() OVER(ORDER BY val) AS rownum
FROM temp;

(795 rows affected)

Total execution time: 00:00:00.025

val rownum 12.50 1 18.40 2 23.80 3 28.00 4 30.00 5 33.75 6 36.00 7 40.00 8 45.00 9 48.00 10 48.75 11 49.80 12 52.35 13 55.20 14 55.80 15 57.50 16 57.80 17 60.00 18 63.00 19 69.60 20 70.00 21 72.00 22 72.96 23 74.40 24 76.00 25 80.10 26 86.40 27 86.50 28 86.85 29 88.00 30 88.50 31 88.80 32 93.50 33 96.50 34 98.40 35 100.80 36 102.40 37 103.20 38 108.00 39 108.50 40 110.00 41 112.00 42 114.00 43 120.00 44 121.60 45 122.40 46 125.00 47 131.75 48 136.00 49 136.30 50 136.80 51 137.50 52 139.80 53 140.00 54 141.60 55 142.50 56 144.00 57 144.80 58 147.00 59 147.90 60 149.00 61 150.00 62 154.40 63 155.00 64 155.40 65 156.00 66 159.00 67 160.00 68 164.40 69 166.00 70 174.90 71 176.10 72 180.40 73 180.48 74 182.40 75 187.00 76 189.42 77 190.00 78 191.10 79 192.00 80 200.00 81 210.00 82 211.20 83 216.00 84 220.00 85 224.00 86 224.83 87 225.50 88 228.00 89 230.40 90 230.85 91 232.09 92 235.20 93 236.25 94 237.90 95 240.00 96 240.10 97 240.40 98 243.18 99 245.00 100 246.24 101 248.00 102 250.80 103 251.50 104 252.00 105 259.51 106 265.35 107 266.00 108 268.80 109 270.00 110 273.60 111 278.00 112 282.00 113 285.12 114 287.80 115 288.00 116 291.55 117 291.84 118 295.38 119 299.25 120 300.00 121 305.00 122 310.00 123 314.76 124 317.75 125 318.84 126 319.20 127 320.00 128 321.12 129 326.00 130 329.69 131 330.00 132 331.20 133 334.80 134 336.00 135 336.80 136 338.00 137 338.20 138 342.00 139 344.00 140 346.56 141 349.50 142 351.00 143 352.00 144 352.60 145 353.20 146 355.50 147 360.00 148 361.00 149 363.60 150 365.89 151 368.93 152 372.38 153 375.50 154 378.00 155 386.20 156 387.50 157 390.00 158 392.20 159 393.00 160 397.80 161 399.00 162 399.85 163 400.00 164 403.20 165 405.75 166 406.40 167 407.70 168 412.35 169 415.80 170 417.20 171 420.00 172 424.00 173 425.12 174 427.50 175 429.40 176 439.00 177 439.20 178 439.60 179 440.00 180 441.15 181 442.00 182 443.40 183 446.60 184 447.20 185 448.00 186 450.00 187 452.90 188 454.00 189 456.00 190 458.76 191 459.00 192 464.00 193 465.70 194 468.45 195 469.11 196 471.20 197 472.50 198 475.11 199 475.15 200 477.00 201 479.40 202 479.75 203 479.80 204 480.00 205 484.26 206 484.50 207 485.00 208 488.70 209 491.50 210 493.00 211 497.52 212 498.10 213 498.50 214 500.00 215 507.00 216 510.00 217 514.40 218 516.00 219 516.47 220 516.80 221 517.44 222 517.80 223 519.00 224 520.41 225 523.26 226 525.00 227 525.30 228 534.85 229 536.40 230 537.50 231 538.60 232 539.40 233 539.50 234 550.00 235 550.59 236 556.62 237 557.60 238 558.00 239 560.00 240 562.60 241 565.50 242 568.80 243 569.00 244 570.00 245 575.00 246 581.00 247 584.00 248 586.00 249 589.00 250 591.60 251 593.75 252 595.50 253 601.83 254 602.40 255 604.22 256 605.00 257 608.00 258 608.40 259 611.30 260 613.20 261 616.00 262 616.50 263 619.50 264 622.35 265 625.00 266 625.28 267 629.00 268 629.50 269 631.60 270 632.40 271 636.00 272 637.50 273 638.50 274 639.90 275 642.00 276 642.06 277 642.20 278 644.80 279 645.00 280 649.00 281 651.00 282 654.06 283 655.00 284 660.00 285 663.10 286 668.80 287 670.00 288 677.00 289 678.00 290 683.30 291 686.00 292 686.70 293 691.20 294 695.63 295 696.00 296 706.00 297 708.75 298 709.55 299 711.00 300 713.30 301 716.72 302 717.50 303 717.60 304 718.08 305 720.00 306 720.90 307 722.50 308 730.00 309 731.50 310 732.60 311 735.00 312 736.00 313 742.50 314 748.80 315 749.06 316 750.50 317 755.00 318 756.00 319 758.50 320 764.30 321 768.75 322 781.00 323 792.75 324 796.35 325 799.20 326 800.00 327 801.10 328 802.00 329 805.43 330 807.38 331 808.00 332 810.00 333 812.50 334 813.37 335 814.42 336 814.50 337 816.30 338 817.88 339 818.40 340 819.00 341 827.55 342 833.00 343 834.20 344 835.20 345 838.45 346 843.00 347 844.25 348 845.80 349 848.00 350 848.70 351 851.20 352 852.00 353 855.02 354 858.00 355 860.10 356 862.50 357 863.28 358 863.43 359 863.60 360 868.75 361 877.20 362 877.73 363 878.00 364 880.50 365 889.70 366 890.00 367 892.64 368 899.00 369 903.60 370 903.75 371 906.93 372 910.40 373 912.00 374 914.40 375 917.00 

### Explanation:

The proposition is asking us to find an alternative way to achieve the distinct value and the row number they are in the table output. The example query uses a group by to get the distinct values so the alternative way I used was using the distinct operator to get the unique values in the val column of the view. and putting that into a cte so that we can represent the val columns as just the unique values. I select the vals from the cte and the row number on the over clause gets the row number for each value to get the desired output in a different way.

# Top 3 - Chapter 7 Q1 - Window functions, Described

- Write a query that calculates the running total value for employee and month
- Tables Involved: Sales.EmpOrders

| empid | ordermonth | val | runval |
| --- | --- | --- | --- |
| 1 | 2014-07-01 | 1614.88 | 1614.88 |
| 1 | 2014-08-01 | 5555.90 | 7170.78 |
| 1 | 2014-09-01 | 6651.00 | 13821.78 |
| 1 | 2014-10-01 | 3933.18 | 17754.96 |
| 1 | 2014-11-01 | 9562.65 | 27317.61 |
| ... |  |  |  |
| 2 | 2014-07-01 | 1176.00 | 1176.00 |
| 2 | 2014-08-01 | 1814.00 | 2990.00 |
| 2 | 2014-09-01 | 2950.80 | 5940.80 |
| 2 | 2014-10-01 | 5164.00 | 11104.80 |
| 2 | 2014-11-01 | 4614.58 | 15719.38 |
| ... |  |  |  |

(192 row(s) affected)

In [10]:
use TSQLV6
SELECT empid, ordermonth, val,
  SUM(val) OVER(PARTITION BY empid
                ORDER BY ordermonth
                ROWS BETWEEN UNBOUNDED PRECEDING
                         AND CURRENT ROW) AS runval
FROM Sales.EmpOrders;

(192 rows affected)

Total execution time: 00:00:00.041

empid ordermonth val runval 1 2020-07-01 1614.88 1614.88 1 2020-08-01 5555.90 7170.78 1 2020-09-01 6651.00 13821.78 1 2020-10-01 3933.18 17754.96 1 2020-11-01 9562.65 27317.61 1 2020-12-01 8446.91 35764.52 1 2021-01-01 7331.60 43096.12 1 2021-02-01 1946.40 45042.52 1 2021-03-01 5124.08 50166.60 1 2021-04-01 240.00 50406.60 1 2021-05-01 9115.96 59522.56 1 2021-06-01 5468.35 64990.91 1 2021-07-01 19530.93 84521.84 1 2021-08-01 5104.70 89626.54 1 2021-09-01 7441.55 97068.09 1 2021-10-01 12414.15 109482.24 1 2021-11-01 4089.90 113572.14 1 2021-12-01 15340.46 128912.60 1 2022-01-01 8115.36 137027.96 1 2022-02-01 11147.51 148175.47 1 2022-03-01 24827.45 173002.92 1 2022-04-01 12587.23 185590.15 1 2022-05-01 6517.47 192107.62 2 2020-07-01 1176.00 1176.00 2 2020-08-01 1814.00 2990.00 2 2020-09-01 2950.80 5940.80 2 2020-10-01 5164.00 11104.80 2 2020-11-01 4614.58 15719.38 2 2020-12-01 6037.68 21757.06 2 2021-01-01 3059.88 24816.94 2 2021-02-01 1584.00 26400.94 2 2021-03-01 2844.90 29245.84 2 2021-04-01 13118.65 42364.49 2 2021-05-01 4373.32 46737.81 2 2021-06-01 6882.20 53620.01 2 2021-07-01 7965.00 61585.01 2 2021-08-01 57.50 61642.51 2 2021-09-01 9286.65 70929.16 2 2021-10-01 10164.80 81093.96 2 2021-11-01 3567.00 84660.96 2 2021-12-01 7540.24 92201.20 2 2022-01-01 4351.11 96552.31 2 2022-02-01 23127.55 119679.86 2 2022-03-01 13937.64 133617.50 2 2022-04-01 30990.28 164607.78 2 2022-05-01 1929.98 166537.76 3 2020-07-01 2963.22 2963.22 3 2020-08-01 3452.08 6415.30 3 2020-09-01 1762.00 8177.30 3 2020-10-01 3579.88 11757.18 3 2020-11-01 3707.98 15465.16 3 2020-12-01 2758.80 18223.96 3 2021-01-01 6981.02 25204.98 3 2021-02-01 10212.64 35417.62 3 2021-03-01 11599.40 47017.02 3 2021-04-01 10252.55 57269.57 3 2021-05-01 18049.60 75319.17 3 2021-06-01 5599.79 80918.96 3 2021-07-01 1081.97 82000.93 3 2021-08-01 5831.60 87832.53 3 2021-09-01 3555.90 91388.43 3 2021-10-01 7626.96 99015.39 3 2021-11-01 9598.08 108613.47 3 2021-12-01 17636.66 126250.13 3 2022-01-01 25705.01 151955.14 3 2022-02-01 21540.24 173495.38 3 2022-03-01 16360.13 189855.51 3 2022-04-01 12957.36 202812.87 4 2020-07-01 11860.45 11860.45 4 2020-08-01 3379.98 15240.43 4 2020-09-01 3575.10 18815.53 4 2020-10-01 13718.97 32534.50 4 2020-11-01 11311.42 43845.92 4 2020-12-01 6099.20 49945.12 4 2021-01-01 23736.47 73681.59 4 2021-02-01 12122.00 85803.59 4 2021-03-01 5230.08 91033.67 4 2021-04-01 13475.99 104509.66 4 2021-05-01 6915.28 111424.94 4 2021-06-01 4082.85 115507.79 4 2021-07-01 5530.90 121038.69 4 2021-08-01 16485.54 137524.23 4 2021-09-01 7931.29 145455.52 4 2021-10-01 10439.84 155895.36 4 2021-11-01 6624.25 162519.61 4 2021-12-01 16235.32 178754.93 4 2022-01-01 19245.97 198000.90 4 2022-02-01 10643.06 208643.96 4 2022-03-01 8298.45 216942.41 4 2022-04-01 9937.71 226880.12 4 2022-05-01 6010.75 232890.87 5 2020-07-01 1638.82 1638.82 5 2020-09-01 1420.00 3058.82 5 2020-10-01 1393.20 4452.02 5 2020-11-01 3901.08 8353.10 5 2020-12-01 10030.82 18383.92 5 2021-03-01 2520.40 20904.32 5 2021-05-01 4500.28 25404.60 5 2021-06-01 3037.40 28442.00 5 2021-07-01 6475.40 34917.40 5 2021-08-01 3585.58 38502.98 5 2021-09-01 2024.83 40527.81 5 2021-10-01 7581.33 48109.14 5 2021-11-01 484.26 48593.40 5 2021-12-01 507.00 49100.40 5 2022-01-01 11702.80 60803.20 5 2022-02-01 5377.06 66180.26 5 2022-03-01 2402.04 68582.30 5 2022-04-01 210.00 68792.30 6 2020-07-01 2559.03 2559.03 6 2020-08-01 2540.12 5099.15 6 2020-09-01 3983.60 9082.75 6 2020-11-01 2228.46 11311.21 6 2020-12-01 5331.40 16642.61 6 2021-01-01 1380.00 18022.61 6 2021-02-01 1324.24 19346.85 6 2021-03-01 1195.20 20542.05 6 2021-04-01 9593.50 30135.55 6 2021-05-01 747.70 30883.25 6 2021-06-01 3464.81 34348.06 6 2021-07-01 1301.00 35649.06 6 2021-08-01 3576.44 39225.50 6 2021-09-01 604.22 39829.72 6 2021-10-01 6185.40 46015.12 6 2021-11-01 6322.45 52337.57 6 2021-12-01 7431.42 59768.99 6 2022-01-01 1874.83 61643.82 6 2022-02-01 1953.40 63597.22 6 2022-03-01 5068.98 68666.20 6 2022-04-01 5246.95 73913.15 7 2020-08-01 479.40 479.40 7 2

### Explanation:

The proposition is asking us to calculate a running total value from the view Sales.EmpOrders based on each employee and month. We select the empid and ordermonth to get the employee and the months. The running total is from a over clause with the sum of the val column partition by the employee's id that is ordered by the month. The running total is used by that with the val of the previous row and the current row to consisntly add to the sum one row at a time.

# Top 4 - Chapter 7 Q7 - Offset Window Functions, Lag and Lead

- Write a query that returns order val of current, previous, and next row from the OrderValues view
- Tables Involved: Sales.OrderValues

| custid | orderid | val | prevval | nextval |
| --- | --- | --- | --- | --- |
| 1 | 10643 | 814.50 | NULL | 878.00 |
| 1 | 10692 | 878.00 | 814.50 | 330.00 |
| 1 | 10702 | 330.00 | 878.00 | 845.80 |
| 1 | 10835 | 845.80 | 330.00 | 471.20 |
| 1 | 10952 | 471.20 | 845.80 | 933.50 |
| 1 | 11011 | 933.50 | 471.20 | NULL |
| 2 | 10308 | 88.80 | NULL | 479.75 |
| 2 | 10625 | 479.75 | 88.80 | 320.00 |
| 2 | 10759 | 320.00 | 479.75 | 514.40 |
| 2 | 10926 | 514.40 | 320.00 | NULL |
| 3 | 10365 | 403.20 | NULL | 749.06 |
| 3 | 10507 | 749.06 | 403.20 | 1940.85 |
| 3 | 10535 | 1940.85 | 749.06 | 2082.00 |
| 3 | 10573 | 2082.00 | 1940.85 | 813.37 |
| 3 | 10677 | 813.37 | 2082.00 | 375.50 |
| 3 | 10682 | 375.50 | 813.37 | 660.00 |
| 3 | 10856 | 660.00 | 375.50 | NULL |
| ... |  |  |  |  |

(830 row(s) affected)

In [11]:
use TSQLV6
SELECT custid, orderid, val,
  LAG(val)  OVER(PARTITION BY custid
                 ORDER BY orderdate, orderid) AS prevval,
  LEAD(val) OVER(PARTITION BY custid
                 ORDER BY orderdate, orderid) AS nextval
FROM Sales.OrderValues
ORDER BY custid, orderdate, orderid;

(830 rows affected)

Total execution time: 00:00:00.012

custid orderid val prevval nextval 1 10643 814.50 NULL 878.00 1 10692 878.00 814.50 330.00 1 10702 330.00 878.00 845.80 1 10835 845.80 330.00 471.20 1 10952 471.20 845.80 933.50 1 11011 933.50 471.20 NULL 2 10308 88.80 NULL 479.75 2 10625 479.75 88.80 320.00 2 10759 320.00 479.75 514.40 2 10926 514.40 320.00 NULL 3 10365 403.20 NULL 749.06 3 10507 749.06 403.20 1940.85 3 10535 1940.85 749.06 2082.00 3 10573 2082.00 1940.85 813.37 3 10677 813.37 2082.00 375.50 3 10682 375.50 813.37 660.00 3 10856 660.00 375.50 NULL 4 10355 480.00 NULL 899.00 4 10383 899.00 480.00 407.70 4 10453 407.70 899.00 2142.90 4 10558 2142.90 407.70 1641.00 4 10707 1641.00 2142.90 228.00 4 10741 228.00 1641.00 319.20 4 10743 319.20 228.00 1477.00 4 10768 1477.00 319.20 191.10 4 10793 191.10 1477.00 282.00 4 10864 282.00 191.10 390.00 4 10920 390.00 282.00 4441.25 4 10953 4441.25 390.00 491.50 4 11016 491.50 4441.25 NULL 5 10278 1488.80 NULL 613.20 5 10280 613.20 1488.80 2222.40 5 10384 2222.40 613.20 1031.70 5 10444 1031.70 2222.40 174.90 5 10445 174.90 1031.70 3192.65 5 10524 3192.65 174.90 1501.09 5 10572 1501.09 3192.65 1503.60 5 10626 1503.60 1501.09 601.83 5 10654 601.83 1503.60 3815.25 5 10672 3815.25 601.83 472.50 5 10689 472.50 3815.25 1459.00 5 10733 1459.00 472.50 96.50 5 10778 96.50 1459.00 1064.50 5 10837 1064.50 96.50 2048.21 5 10857 2048.21 1064.50 1096.20 5 10866 1096.20 2048.21 709.55 5 10875 709.55 1096.20 1835.70 5 10924 1835.70 709.55 NULL 6 10501 149.00 NULL 136.80 6 10509 136.80 149.00 330.00 6 10582 330.00 136.80 464.00 6 10614 464.00 330.00 625.00 6 10853 625.00 464.00 677.00 6 10956 677.00 625.00 858.00 6 11058 858.00 677.00 NULL 7 10265 1176.00 NULL 1420.00 7 10297 1420.00 1176.00 7390.20 7 10360 7390.20 1420.00 1994.52 7 10436 1994.52 7390.20 1838.20 7 10449 1838.20 1994.52 520.41 7 10559 520.41 1838.20 1761.00 7 10566 1761.00 520.41 593.75 7 10584 593.75 1761.00 450.00 7 10628 450.00 593.75 660.00 7 10679 660.00 450.00 730.00 7 10826 730.00 660.00 NULL 8 10326 982.00 NULL 3026.85 8 10801 3026.85 982.00 224.00 8 10970 224.00 3026.85 NULL 9 10331 88.50 NULL 2436.18 9 10340 2436.18 88.50 1549.60 9 10362 1549.60 2436.18 1820.80 9 10470 1820.80 1549.60 2550.00 9 10511 2550.00 1820.80 818.40 9 10525 818.40 2550.00 1930.40 9 10663 1930.40 818.40 1296.00 9 10715 1296.00 1930.40 484.26 9 10730 484.26 1296.00 360.00 9 10732 360.00 484.26 1948.50 9 10755 1948.50 360.00 843.00 9 10827 843.00 1948.50 1979.23 9 10871 1979.23 843.00 917.00 9 10876 917.00 1979.23 1788.63 9 10932 1788.63 917.00 360.00 9 10940 360.00 1788.63 792.75 9 11076 792.75 360.00 NULL 10 10389 1832.80 NULL 802.00 10 10410 802.00 1832.80 966.80 10 10411 966.80 802.00 1892.25 10 10431 1892.25 966.80 851.20 10 10492 851.20 1892.25 3118.00 10 10742 3118.00 851.20 1447.50 10 10918 1447.50 3118.00 1025.33 10 10944 1025.33 1447.50 4422.00 10 10949 4422.00 1025.33 717.50 10 10975 717.50 4422.00 1014.00 10 10982 1014.00 717.50 877.73 10 11027 877.73 1014.00 1309.50 10 11045 1309.50 877.73 525.00 10 11048 525.00 1309.50 NULL 11 10289 479.40 NULL 1328.00 11 10471 1328.00 479.40 386.20 11 10484 386.20 1328.00 139.80 11 10538 139.80 386.20 355.50 11 10539 355.50 139.80 477.00 11 10578 477.00 355.50 493.00 11 10599 493.00 477.00 711.00 11 10943 711.00 493.00 220.00 11 10947 220.00 711.00 1500.00 11 11023 1500.00 220.00 NULL 12 10521 225.50 NULL 12.50 12 10782 12.50 225.50 477.00 12 10819 477.00 12.50 150.00 12 10881 150.00 477.00 644.80 12 10937 644.80 150.00 305.00 12 11054 305.00 644.80 NULL 13 10259 100.80 NULL NULL 14 10254 556.62 NULL 1117.60 14 10370 1117.60 556.62 2314.20 14 10519 2314.20 1117.60 1890.50 14 10731 1890.50 2314.20 2311.70 14 10746 2311.70 1890.50 1098.46 14 10966 1098.46 2311.70 1286.80 14 11029 1286.80 1098.46 1773.00 14 11041 1773.00 1286.80 NULL 15 10290 2169.00 NULL 216.00 15 10466 216.00 2169.00 912.00 15 10494 912.00 216.00 108.00 15 10969 108.00 912.00 405.75 15 11042 405.75 108.00 NULL 16 10435 631.60 NULL 156.00 16 10462 156.00 631.60 931.50 16 10848 931.5

### Explanation:

The proposition is asking us to use the view, Sales.OrderValues to get the order val of the current, previous, and next val order. I start by selecting the custid, orderid, and val for the order and the next column finds the previous val number. Using the lag offset function taking the val column and adding a over clause to a partition by on the custid and ordered by the date and id of the orders. Repeat the syntax excpet switch lag with lead to get the next val and it creates the desired output.

# Top 5 - Chapter 7 Q8- Offset Window Functions, First\_Value and Last\_Value

- Write a query that returns the value of the current, first, and last orders for each customer
- Tables Involved: Sales.OrderValues

| custid | orderid | val | firstval | lastval |
| --- | --- | --- | --- | --- |
| 1 | 10643 | 814.50 | 814.50 | 933.50 |
| 1 | 10692 | 878.00 | 814.50 | 933.50 |
| 1 | 10702 | 330.00 | 814.50 | 933.50 |
| 1 | 10835 | 845.80 | 814.50 | 933.50 |
| 1 | 10952 | 471.20 | 814.50 | 933.50 |
| 1 | 11011 | 933.50 | 814.50 | 933.50 |
| 2 | 10308 | 88.80 | 88.80 | 514.40 |
| 2 | 10625 | 479.75 | 88.80 | 514.40 |
| 2 | 10759 | 320.00 | 88.80 | 514.40 |
| 2 | 10926 | 514.40 | 88.80 | 514.40 |
| 3 | 10365 | 403.20 | 403.20 | 660.00 |
| 3 | 10507 | 749.06 | 403.20 | 660.00 |
| 3 | 10535 | 1940.85 | 403.20 | 660.00 |
| 3 | 10573 | 2082.00 | 403.20 | 660.00 |
| 3 | 10677 | 813.37 | 403.20 | 660.00 |
| 3 | 10682 | 375.50 | 403.20 | 660.00 |
| 3 | 10856 | 660.00 | 403.20 | 660.00 |
| ... |  |  |  |  |

(830 row(s) affected)

In [12]:
use TSQLV6
SELECT custid, orderid, val,
  FIRST_VALUE(val) OVER(PARTITION BY custid
                        ORDER BY orderdate, orderid
                        ROWS BETWEEN UNBOUNDED PRECEDING
                                 AND CURRENT ROW) AS firstval,
  LAST_VALUE(val)  OVER(PARTITION BY custid
                        ORDER BY orderdate, orderid
                        ROWS BETWEEN CURRENT ROW
                                 AND UNBOUNDED FOLLOWING) AS lastval
FROM Sales.OrderValues
ORDER BY custid, orderdate, orderid;

(830 rows affected)

Total execution time: 00:00:00.037

custid orderid val firstval lastval 1 10643 814.50 814.50 933.50 1 10692 878.00 814.50 933.50 1 10702 330.00 814.50 933.50 1 10835 845.80 814.50 933.50 1 10952 471.20 814.50 933.50 1 11011 933.50 814.50 933.50 2 10308 88.80 88.80 514.40 2 10625 479.75 88.80 514.40 2 10759 320.00 88.80 514.40 2 10926 514.40 88.80 514.40 3 10365 403.20 403.20 660.00 3 10507 749.06 403.20 660.00 3 10535 1940.85 403.20 660.00 3 10573 2082.00 403.20 660.00 3 10677 813.37 403.20 660.00 3 10682 375.50 403.20 660.00 3 10856 660.00 403.20 660.00 4 10355 480.00 480.00 491.50 4 10383 899.00 480.00 491.50 4 10453 407.70 480.00 491.50 4 10558 2142.90 480.00 491.50 4 10707 1641.00 480.00 491.50 4 10741 228.00 480.00 491.50 4 10743 319.20 480.00 491.50 4 10768 1477.00 480.00 491.50 4 10793 191.10 480.00 491.50 4 10864 282.00 480.00 491.50 4 10920 390.00 480.00 491.50 4 10953 4441.25 480.00 491.50 4 11016 491.50 480.00 491.50 5 10278 1488.80 1488.80 1835.70 5 10280 613.20 1488.80 1835.70 5 10384 2222.40 1488.80 1835.70 5 10444 1031.70 1488.80 1835.70 5 10445 174.90 1488.80 1835.70 5 10524 3192.65 1488.80 1835.70 5 10572 1501.09 1488.80 1835.70 5 10626 1503.60 1488.80 1835.70 5 10654 601.83 1488.80 1835.70 5 10672 3815.25 1488.80 1835.70 5 10689 472.50 1488.80 1835.70 5 10733 1459.00 1488.80 1835.70 5 10778 96.50 1488.80 1835.70 5 10837 1064.50 1488.80 1835.70 5 10857 2048.21 1488.80 1835.70 5 10866 1096.20 1488.80 1835.70 5 10875 709.55 1488.80 1835.70 5 10924 1835.70 1488.80 1835.70 6 10501 149.00 149.00 858.00 6 10509 136.80 149.00 858.00 6 10582 330.00 149.00 858.00 6 10614 464.00 149.00 858.00 6 10853 625.00 149.00 858.00 6 10956 677.00 149.00 858.00 6 11058 858.00 149.00 858.00 7 10265 1176.00 1176.00 730.00 7 10297 1420.00 1176.00 730.00 7 10360 7390.20 1176.00 730.00 7 10436 1994.52 1176.00 730.00 7 10449 1838.20 1176.00 730.00 7 10559 520.41 1176.00 730.00 7 10566 1761.00 1176.00 730.00 7 10584 593.75 1176.00 730.00 7 10628 450.00 1176.00 730.00 7 10679 660.00 1176.00 730.00 7 10826 730.00 1176.00 730.00 8 10326 982.00 982.00 224.00 8 10801 3026.85 982.00 224.00 8 10970 224.00 982.00 224.00 9 10331 88.50 88.50 792.75 9 10340 2436.18 88.50 792.75 9 10362 1549.60 88.50 792.75 9 10470 1820.80 88.50 792.75 9 10511 2550.00 88.50 792.75 9 10525 818.40 88.50 792.75 9 10663 1930.40 88.50 792.75 9 10715 1296.00 88.50 792.75 9 10730 484.26 88.50 792.75 9 10732 360.00 88.50 792.75 9 10755 1948.50 88.50 792.75 9 10827 843.00 88.50 792.75 9 10871 1979.23 88.50 792.75 9 10876 917.00 88.50 792.75 9 10932 1788.63 88.50 792.75 9 10940 360.00 88.50 792.75 9 11076 792.75 88.50 792.75 10 10389 1832.80 1832.80 525.00 10 10410 802.00 1832.80 525.00 10 10411 966.80 1832.80 525.00 10 10431 1892.25 1832.80 525.00 10 10492 851.20 1832.80 525.00 10 10742 3118.00 1832.80 525.00 10 10918 1447.50 1832.80 525.00 10 10944 1025.33 1832.80 525.00 10 10949 4422.00 1832.80 525.00 10 10975 717.50 1832.80 525.00 10 10982 1014.00 1832.80 525.00 10 11027 877.73 1832.80 525.00 10 11045 1309.50 1832.80 525.00 10 11048 525.00 1832.80 525.00 11 10289 479.40 479.40 1500.00 11 10471 1328.00 479.40 1500.00 11 10484 386.20 479.40 1500.00 11 10538 139.80 479.40 1500.00 11 10539 355.50 479.40 1500.00 11 10578 477.00 479.40 1500.00 11 10599 493.00 479.40 1500.00 11 10943 711.00 479.40 1500.00 11 10947 220.00 479.40 1500.00 11 11023 1500.00 479.40 1500.00 12 10521 225.50 225.50 305.00 12 10782 12.50 225.50 305.00 12 10819 477.00 225.50 305.00 12 10881 150.00 225.50 305.00 12 10937 644.80 225.50 305.00 12 11054 305.00 225.50 305.00 13 10259 100.80 100.80 100.80 14 10254 556.62 556.62 1773.00 14 10370 1117.60 556.62 1773.00 14 10519 2314.20 556.62 1773.00 14 10731 1890.50 556.62 1773.00 14 10746 2311.70 556.62 1773.00 14 10966 1098.46 556.62 1773.00 14 11029 1286.80 556.62 1773.00 14 11041 1773.00 556.62 1773.00 15 10290 2169.00 2169.00 405.75 15 10466 216.00 2169.00 405.75 15 10494 912.00 2169.00 405.75 15 10969 108.00 2169.00 405.75 15 11042 405.75 2169.00 405.75 16 10435 631.60 631.60 931.50 16 10462 

### Explanation:

The proposition is asking us to use the view, Sales.OrderValues to get the order val of the current, first, and last val order for each customer. I start by selecting the custid, orderid, and val for the order and the next column finds the value of the first order for that customer. Using the First Value offset function taking the val column and adding a over clause to a partition by on the custid and ordered by the date and id of the orders. Then we add a Rows Between clause to filter the rows from a unbounded point to the current row. Repeat the syntax excpet switch first with last to get the last value with the rows between looking from the current row and no upper bound and it creates the desired output.

# Chapter 7 Q2 - Ranking Window functions

- Write a query that ranks the values using all the ranking functions
- Tables Involved: Sales.OrderValues

In [13]:
Use TSQLV6
SELECT orderid, custid, val,
  ROW_NUMBER() OVER(ORDER BY val) AS rownum,
  RANK()       OVER(ORDER BY val) AS rank,
  DENSE_RANK() OVER(ORDER BY val) AS dense_rank,
  NTILE(10)    OVER(ORDER BY val) AS ntile
FROM Sales.OrderValues
ORDER BY val;

(830 rows affected)

Total execution time: 00:00:00.088

orderid custid val rownum rank dense_rank ntile 10782 12 12.50 1 1 1 1 10807 27 18.40 2 2 2 1 10586 66 23.80 3 3 3 1 10767 76 28.00 4 4 4 1 10898 54 30.00 5 5 5 1 10900 88 33.75 6 6 6 1 10883 48 36.00 7 7 7 1 11051 41 36.00 8 7 7 1 10815 71 40.00 9 9 8 1 10674 38 45.00 10 10 9 1 11057 53 45.00 11 10 9 1 10271 75 48.00 12 12 10 1 10602 83 48.75 13 13 11 1 10422 27 49.80 14 14 12 1 10738 74 52.35 15 15 13 1 10754 49 55.20 16 16 14 1 10631 41 55.80 17 17 15 1 10620 42 57.50 18 18 16 1 10963 28 57.80 19 19 17 1 11037 30 60.00 20 20 18 1 10683 18 63.00 21 21 19 1 10992 77 69.60 22 22 20 1 10887 29 70.00 23 23 21 1 10589 32 72.00 24 24 22 1 10371 41 72.96 25 25 23 1 10955 24 74.40 26 26 24 1 11019 64 76.00 27 27 25 1 10288 66 80.10 28 28 26 1 10391 17 86.40 29 29 27 1 10281 69 86.50 30 30 28 1 11067 17 86.85 31 31 29 1 10753 27 88.00 32 32 30 1 10331 9 88.50 33 33 31 1 10308 2 88.80 34 34 32 1 10710 27 93.50 35 35 33 1 10778 5 96.50 36 36 34 1 10867 48 98.40 37 37 35 1 10259 13 100.80 38 38 36 1 10415 36 102.40 39 39 37 1 10378 24 103.20 40 40 38 1 10969 15 108.00 41 41 39 1 10907 74 108.50 42 42 40 1 10950 49 110.00 43 43 41 1 10531 54 110.00 44 43 41 1 10381 46 112.00 45 45 42 1 10322 58 112.00 46 45 42 1 10699 52 114.00 47 47 43 1 10615 90 120.00 48 48 44 1 10295 85 121.60 49 49 45 1 10899 46 122.40 50 50 46 1 10662 48 125.00 51 51 47 1 10959 31 131.75 52 52 48 1 10366 29 136.00 53 53 49 1 10352 28 136.30 54 54 50 1 10509 6 136.80 55 55 51 1 10928 29 137.50 56 56 52 1 10538 11 139.80 57 57 53 1 10737 85 139.80 58 57 53 1 10809 88 140.00 59 59 54 1 10349 75 141.60 60 60 55 1 10585 88 142.50 61 61 56 1 10321 38 144.00 62 62 57 1 10334 84 144.80 63 63 58 1 10482 43 147.00 64 64 59 1 10505 51 147.90 65 65 60 1 10501 6 149.00 66 66 61 1 10881 12 150.00 67 67 62 1 10655 66 154.40 68 68 63 1 10568 29 155.00 69 69 64 1 10282 69 155.40 70 70 65 1 10462 16 156.00 71 71 66 1 10843 84 159.00 72 72 67 1 10870 91 160.00 73 73 68 1 10323 39 164.40 74 74 69 1 10386 21 166.00 75 75 70 1 10445 5 174.90 76 76 71 1 10460 24 176.10 77 77 72 1 10708 77 180.40 78 78 73 1 10476 35 180.48 79 79 74 1 10313 63 182.40 80 80 75 1 10810 42 187.00 81 81 76 1 11065 46 189.42 82 82 77 1 10496 81 190.00 83 83 78 1 10793 4 191.10 84 84 79 2 10428 66 192.00 85 85 80 2 10520 70 200.00 86 86 81 2 11040 32 200.00 87 86 81 2 11043 74 210.00 88 88 82 2 10545 43 210.00 89 88 82 2 10840 47 211.20 90 90 83 2 10466 15 216.00 91 91 84 2 10947 11 220.00 92 92 85 2 10970 8 224.00 93 93 86 2 10777 31 224.00 94 93 86 2 10414 21 224.83 95 95 87 2 10521 12 225.50 96 96 88 2 10775 78 228.00 97 97 89 2 10741 4 228.00 98 97 89 2 10473 38 230.40 99 99 90 2 10604 28 230.85 100 100 91 2 11074 73 232.09 101 101 92 2 10467 49 235.20 102 102 93 2 10770 34 236.25 103 103 94 2 10822 82 237.90 104 104 95 2 10739 85 240.00 105 105 96 2 10508 56 240.00 106 105 96 2 10548 79 240.10 107 107 97 2 10318 38 240.40 108 108 98 2 11014 47 243.18 109 109 99 2 10945 52 245.00 110 110 100 2 10446 79 246.24 111 111 101 2 10980 24 248.00 112 112 102 2 10824 24 250.80 113 113 103 2 10972 40 251.50 114 114 104 2 10752 53 252.00 115 115 105 2 10491 28 259.51 116 116 106 2 10960 35 265.35 117 117 107 2 11060 27 266.00 118 118 108 2 10311 18 268.80 119 119 109 2 11025 87 270.00 120 120 110 2 11049 31 273.60 121 121 111 2 10495 42 278.00 122 122 112 2 10864 4 282.00 123 123 113 2 10336 60 285.12 124 124 114 2 10725 21 287.80 125 125 115 2 10317 48 288.00 126 126 116 2 10973 40 291.55 127 127 117 2 10275 49 291.84 128 128 118 2 11004 50 295.38 129 129 119 2 10610 41 299.25 130 130 120 2 11073 58 300.00 131 131 121 2 11054 12 305.00 132 132 122 2 10874 30 310.00 133 133 123 2 10581 21 310.00 134 133 123 2 10794 61 314.76 135 135 124 2 10579 45 317.75 136 136 125 2 10652 31 318.84 137 137 126 2 10743 4 319.20 138 138 127 2 10409 54 319.20 139 138 127 2 10759 2 320.00 140 140 128 2 10434 24 321.12 141 141 129 2 11003 78 326.00 142 142 130 2 11006 32 329.69 143 143 131 2 10702 1 330.00 144 144 132 2 10582 6 330.00 145

### Explanation:

We need to use all the ranking functions available to rank the values from the view Sales.OrderValues, so we just use the ranking functions row\_number, rank, dense\_rank, and ntile with all having the same over clause that is ordered by the value. The only difference for this specific case is the ntile function that breaks up the rows into 10 groups and evenly assigning the values so that there is about the same number of rows with a rank of each number from 1-10.

# Chapter 7 Q3 - Ranking Window functions

- Write a query that numbers each order for each customer
- Tables Involved: Sales.OrderValues

In [14]:
use TSQLV6
SELECT orderid, custid, val,
  ROW_NUMBER() OVER(PARTITION BY custid
                    ORDER BY val) AS rownum
FROM Sales.OrderValues
ORDER BY custid, val;

(830 rows affected)

Total execution time: 00:00:00.034

orderid custid val rownum 10702 1 330.00 1 10952 1 471.20 2 10643 1 814.50 3 10835 1 845.80 4 10692 1 878.00 5 11011 1 933.50 6 10308 2 88.80 1 10759 2 320.00 2 10625 2 479.75 3 10926 2 514.40 4 10682 3 375.50 1 10365 3 403.20 2 10856 3 660.00 3 10507 3 749.06 4 10677 3 813.37 5 10535 3 1940.85 6 10573 3 2082.00 7 10793 4 191.10 1 10741 4 228.00 2 10864 4 282.00 3 10743 4 319.20 4 10920 4 390.00 5 10453 4 407.70 6 10355 4 480.00 7 11016 4 491.50 8 10383 4 899.00 9 10768 4 1477.00 10 10707 4 1641.00 11 10558 4 2142.90 12 10953 4 4441.25 13 10778 5 96.50 1 10445 5 174.90 2 10689 5 472.50 3 10654 5 601.83 4 10280 5 613.20 5 10875 5 709.55 6 10444 5 1031.70 7 10837 5 1064.50 8 10866 5 1096.20 9 10733 5 1459.00 10 10278 5 1488.80 11 10572 5 1501.09 12 10626 5 1503.60 13 10924 5 1835.70 14 10857 5 2048.21 15 10384 5 2222.40 16 10524 5 3192.65 17 10672 5 3815.25 18 10509 6 136.80 1 10501 6 149.00 2 10582 6 330.00 3 10614 6 464.00 4 10853 6 625.00 5 10956 6 677.00 6 11058 6 858.00 7 10628 7 450.00 1 10559 7 520.41 2 10584 7 593.75 3 10679 7 660.00 4 10826 7 730.00 5 10265 7 1176.00 6 10297 7 1420.00 7 10566 7 1761.00 8 10449 7 1838.20 9 10436 7 1994.52 10 10360 7 7390.20 11 10970 8 224.00 1 10326 8 982.00 2 10801 8 3026.85 3 10331 9 88.50 1 10732 9 360.00 2 10940 9 360.00 3 10730 9 484.26 4 11076 9 792.75 5 10525 9 818.40 6 10827 9 843.00 7 10876 9 917.00 8 10715 9 1296.00 9 10362 9 1549.60 10 10932 9 1788.63 11 10470 9 1820.80 12 10663 9 1930.40 13 10755 9 1948.50 14 10871 9 1979.23 15 10340 9 2436.18 16 10511 9 2550.00 17 11048 10 525.00 1 10975 10 717.50 2 10410 10 802.00 3 10492 10 851.20 4 11027 10 877.73 5 10411 10 966.80 6 10982 10 1014.00 7 10944 10 1025.33 8 11045 10 1309.50 9 10918 10 1447.50 10 10389 10 1832.80 11 10431 10 1892.25 12 10742 10 3118.00 13 10949 10 4422.00 14 10538 11 139.80 1 10947 11 220.00 2 10539 11 355.50 3 10484 11 386.20 4 10578 11 477.00 5 10289 11 479.40 6 10599 11 493.00 7 10943 11 711.00 8 10471 11 1328.00 9 11023 11 1500.00 10 10782 12 12.50 1 10881 12 150.00 2 10521 12 225.50 3 11054 12 305.00 4 10819 12 477.00 5 10937 12 644.80 6 10259 13 100.80 1 10254 14 556.62 1 10966 14 1098.46 2 10370 14 1117.60 3 11029 14 1286.80 4 11041 14 1773.00 5 10731 14 1890.50 6 10746 14 2311.70 7 10519 14 2314.20 8 10969 15 108.00 1 10466 15 216.00 2 11042 15 405.75 3 10494 15 912.00 4 10290 15 2169.00 5 10462 16 156.00 1 10435 16 631.60 2 10848 16 931.50 3 10391 17 86.40 1 11067 17 86.85 2 10797 17 420.00 3 10363 17 447.20 4 10825 17 1030.76 5 11036 17 1692.00 6 10683 18 63.00 1 10311 18 268.80 2 10609 18 424.00 3 10890 18 860.10 4 10726 19 655.00 1 10532 19 796.35 2 11047 19 817.88 3 10364 19 950.00 4 11024 19 1966.81 5 10987 19 2772.00 6 10400 19 3063.00 7 11056 19 3740.00 8 10771 20 344.00 1 10571 20 550.59 2 10403 20 855.02 3 10968 20 1408.00 4 10667 20 1536.80 5 10258 20 1614.88 6 10368 20 1689.78 7 10442 20 1792.00 8 10263 20 1873.80 9 10773 20 2030.40 10 10390 20 2090.88 11 10795 20 2158.00 12 10764 20 2286.00 13 10402 20 2713.50 14 10382 20 2900.00 15 10854 20 2966.50 16 10698 20 3436.44 17 10990 20 4288.85 18 11008 20 4680.90 19 10836 20 4705.50 20 10595 20 4725.00 21 10979 20 4813.50 22 10430 20 4899.20 23 11072 20 5218.00 24 10351 20 5398.73 25 10633 20 5510.59 26 10895 20 6379.40 27 10776 20 6635.28 28 11017 20 6750.00 29 10514 20 8623.45 30 10386 21 166.00 1 10414 21 224.83 2 10725 21 287.80 3 10581 21 310.00 4 10512 21 525.30 5 10347 21 814.42 6 10650 21 1779.20 7 10763 23 616.00 1 10480 23 756.00 2 10408 23 1622.40 3 10789 23 3687.00 4 10634 23 4985.50 5 10955 24 74.40 1 10378 24 103.20 2 10460 24 176.10 3 10980 24 248.00 4 10824 24 250.80 5 10434 24 321.12 6 10264 24 695.63 7 11050 24 810.00 8 10902 24 863.43 9 10774 24 868.75 10 10880 24 1500.00 11 10327 24 1810.00 12 10533 24 2222.20 13 10977 24 2233.00 14 10703 24 2545.00 15 11001 24 2769.00 16 10561 24 2844.50 17 10762 24 4337.00 18 10993 24 4895.44 19 10560 25 1072.43 1 10859 25 1078.69 2 10653 25 1083.15 3 10929 25 1174.75 4 10717 25 1270.75 5

### Explanation:

We get the regular information of the order with the cutsomer's id, the order's id, and the value of the order, and to number each order for the customer we use the row-number function with a over clause that partitions the customer id and orders by the val to assign a rank to each order from lowest to highest.

# Chapter 7 Q4 & 5 - Ranking Window functions

- Write a query that ranks each unique value number

```
SELECT DISTINCT val, ROW_NUMBER() OVER(ORDER BY val) AS rownum
FROM Sales.OrderValues;
```

- Why does this attempt not produce the correct output and what is way achieve the desired output
- Tables Involved: Sales.OrderValues

In [15]:
use TSQLV6
SELECT val, ROW_NUMBER() OVER(ORDER BY val) AS rownum
FROM Sales.OrderValues
GROUP BY val;

(795 rows affected)

Total execution time: 00:00:00.035

val rownum 12.50 1 18.40 2 23.80 3 28.00 4 30.00 5 33.75 6 36.00 7 40.00 8 45.00 9 48.00 10 48.75 11 49.80 12 52.35 13 55.20 14 55.80 15 57.50 16 57.80 17 60.00 18 63.00 19 69.60 20 70.00 21 72.00 22 72.96 23 74.40 24 76.00 25 80.10 26 86.40 27 86.50 28 86.85 29 88.00 30 88.50 31 88.80 32 93.50 33 96.50 34 98.40 35 100.80 36 102.40 37 103.20 38 108.00 39 108.50 40 110.00 41 112.00 42 114.00 43 120.00 44 121.60 45 122.40 46 125.00 47 131.75 48 136.00 49 136.30 50 136.80 51 137.50 52 139.80 53 140.00 54 141.60 55 142.50 56 144.00 57 144.80 58 147.00 59 147.90 60 149.00 61 150.00 62 154.40 63 155.00 64 155.40 65 156.00 66 159.00 67 160.00 68 164.40 69 166.00 70 174.90 71 176.10 72 180.40 73 180.48 74 182.40 75 187.00 76 189.42 77 190.00 78 191.10 79 192.00 80 200.00 81 210.00 82 211.20 83 216.00 84 220.00 85 224.00 86 224.83 87 225.50 88 228.00 89 230.40 90 230.85 91 232.09 92 235.20 93 236.25 94 237.90 95 240.00 96 240.10 97 240.40 98 243.18 99 245.00 100 246.24 101 248.00 102 250.80 103 251.50 104 252.00 105 259.51 106 265.35 107 266.00 108 268.80 109 270.00 110 273.60 111 278.00 112 282.00 113 285.12 114 287.80 115 288.00 116 291.55 117 291.84 118 295.38 119 299.25 120 300.00 121 305.00 122 310.00 123 314.76 124 317.75 125 318.84 126 319.20 127 320.00 128 321.12 129 326.00 130 329.69 131 330.00 132 331.20 133 334.80 134 336.00 135 336.80 136 338.00 137 338.20 138 342.00 139 344.00 140 346.56 141 349.50 142 351.00 143 352.00 144 352.60 145 353.20 146 355.50 147 360.00 148 361.00 149 363.60 150 365.89 151 368.93 152 372.38 153 375.50 154 378.00 155 386.20 156 387.50 157 390.00 158 392.20 159 393.00 160 397.80 161 399.00 162 399.85 163 400.00 164 403.20 165 405.75 166 406.40 167 407.70 168 412.35 169 415.80 170 417.20 171 420.00 172 424.00 173 425.12 174 427.50 175 429.40 176 439.00 177 439.20 178 439.60 179 440.00 180 441.15 181 442.00 182 443.40 183 446.60 184 447.20 185 448.00 186 450.00 187 452.90 188 454.00 189 456.00 190 458.76 191 459.00 192 464.00 193 465.70 194 468.45 195 469.11 196 471.20 197 472.50 198 475.11 199 475.15 200 477.00 201 479.40 202 479.75 203 479.80 204 480.00 205 484.26 206 484.50 207 485.00 208 488.70 209 491.50 210 493.00 211 497.52 212 498.10 213 498.50 214 500.00 215 507.00 216 510.00 217 514.40 218 516.00 219 516.47 220 516.80 221 517.44 222 517.80 223 519.00 224 520.41 225 523.26 226 525.00 227 525.30 228 534.85 229 536.40 230 537.50 231 538.60 232 539.40 233 539.50 234 550.00 235 550.59 236 556.62 237 557.60 238 558.00 239 560.00 240 562.60 241 565.50 242 568.80 243 569.00 244 570.00 245 575.00 246 581.00 247 584.00 248 586.00 249 589.00 250 591.60 251 593.75 252 595.50 253 601.83 254 602.40 255 604.22 256 605.00 257 608.00 258 608.40 259 611.30 260 613.20 261 616.00 262 616.50 263 619.50 264 622.35 265 625.00 266 625.28 267 629.00 268 629.50 269 631.60 270 632.40 271 636.00 272 637.50 273 638.50 274 639.90 275 642.00 276 642.06 277 642.20 278 644.80 279 645.00 280 649.00 281 651.00 282 654.06 283 655.00 284 660.00 285 663.10 286 668.80 287 670.00 288 677.00 289 678.00 290 683.30 291 686.00 292 686.70 293 691.20 294 695.63 295 696.00 296 706.00 297 708.75 298 709.55 299 711.00 300 713.30 301 716.72 302 717.50 303 717.60 304 718.08 305 720.00 306 720.90 307 722.50 308 730.00 309 731.50 310 732.60 311 735.00 312 736.00 313 742.50 314 748.80 315 749.06 316 750.50 317 755.00 318 756.00 319 758.50 320 764.30 321 768.75 322 781.00 323 792.75 324 796.35 325 799.20 326 800.00 327 801.10 328 802.00 329 805.43 330 807.38 331 808.00 332 810.00 333 812.50 334 813.37 335 814.42 336 814.50 337 816.30 338 817.88 339 818.40 340 819.00 341 827.55 342 833.00 343 834.20 344 835.20 345 838.45 346 843.00 347 844.25 348 845.80 349 848.00 350 848.70 351 851.20 352 852.00 353 855.02 354 858.00 355 860.10 356 862.50 357 863.28 358 863.43 359 863.60 360 868.75 361 877.20 362 877.73 363 878.00 364 880.50 365 889.70 366 890.00 367 892.64 368 899.00 369 903.60 370 903.75 371 906.93 372 910.40 373 912.00 374 914.40 375 917.00 

### Explanation:

The first attempt at answering the proposition doesn't work becuase the row\_number clause is proccssed before the distinct clause so it already has ranked each value making it unique and the distinct will not see any duplicates. Fixing this we instead use group by in place of distinct so that we can have one single use of each value and then numbering the rows to get no duplicate value numbers.